src/dotnet/DotNetTopLevelCommandBase.cs

```diff
@@ -3,57 +3,80 @@
 
 using System;
 using System.Collections.Generic;
+using System.IO;
+using System.Linq;
 using Microsoft.DotNet.Cli.CommandLine;
 using Microsoft.DotNet.Cli.Utils;
 using Microsoft.DotNet.Tools;
-using Microsoft.DotNet.Tools.Add.ProjectToProjectReference;
+using Microsoft.DotNet.Tools.Common;
 
 namespace Microsoft.DotNet.Cli
 {
     public abstract class DotNetTopLevelCommandBase
     {
         protected abstract string CommandName { get; }
         protected abstract string FullCommandNameLocalized { get; }
-        internal abstract List<Func<CommandLineApplication, CommandLineApplication>> SubCommands { get; }
+        internal abstract List<Func<DotNetSubCommandBase>> SubCommands { get; }
 
         public int RunCommand(string[] args)
         {
             DebugHelper.HandleDebugSwitch(ref args);
 
-            CommandLineApplication app = new CommandLineApplication(throwOnUnexpectedArg: true)
+            CommandLineApplication command = new CommandLineApplication(throwOnUnexpectedArg: true)
             {
                 Name = $"dotnet {CommandName}",
                 FullName = FullCommandNameLocalized,
             };
 
-            app.HelpOption("-h|--help");
+            command.HelpOption("-h|--help");
 
-            app.Argument(
+            command.Argument(
                 Constants.ProjectOrSolutionArgumentName,
                 CommonLocalizableStrings.ArgumentsProjectOrSolutionDescription);
 
             foreach (var subCommandCreator in SubCommands)
             {
-                subCommandCreator(app);
+                var subCommand = subCommandCreator();
+                command.Command(subCommand);
```

Comment:
The purpose of the `Command` method is a bit unclear. This line of code shows the problem by using the same word for subject, object, and verb. Can we come up with some names that clarify the intent?